core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -156,7 +157,7 @@ class ReplicaFetcherThread(name: String,
 
       warn(s"Reset fetch offset for partition $topicPartition from ${replica.logEndOffset.messageOffset} to current " +
         s"leader's latest offset $leaderEndOffset")
-      replicaMgr.logManager.truncateTo(Map(topicPartition -> leaderEndOffset))
+      replicaMgr.logManager.truncateTo(Map(topicPartition -> leaderEndOffset), isFuture = false)
```

Comment:
Currently the the future replica will only be written by the ReplicaAlterDirThread. And the current log will only be written by ReplicaFetcherThread and KafkaRequestHandler. The ReplicaAlterDirThread is responsible for fetch data from current log to the future log in the same way that ReplicaFetcherThread fetches data from leader to follower. 

The advantage of this design is that 1) the design is simpler because ReplicaAlterDirThread uses essentially the same logic as ReplicaFetcherThread; and 2) most new logic for intra-broker replication is handled by ReplicaAlterDirThread and ReplicaFetcherThread won't need to worry about the future log.

In order to address this problem cleanly while still keep the pattern described above, I have updated the patch so that logManager.truncateTo() will always be called from partition.truncateTo(). And the same for truncateFullyAndStartAt(). In partition.truncateTo() will grab the read lock `partition.leaderIsrUpdateLock` before truncating the log.

With the use of partition.leaderIsrUpdateLock, the current patch ensures that when `partition.maybeDeleteAndSwapFutureReplica()` calls `logManager.deleteAndSwapFutureLog`, it is guaranteed that `replica.logEndOffset == futureReplica.logEndOffset` and no thread can truncate or write the log because all write operation will need to grab `leaderIsrUpdateLock`. Therefore, if the ReplicaFetcherThread truncates the log to an offset smaller than the log end offset of the future log, it is guaranteed that ReplicaAlterDirThread() will receives OffsetOutofRangeException and truncate the future log as well.

Does this make sense?

Reply:
Hmm, does it guarantee that? After the current replica truncates the data, new data could have been written to the current replica. When the future replica fetches from the current replica again, it may have seen the new data and therefore won't receive OffsetOutofRangeException. This will potentially lead to inconsistent data between the future and the current replica.

Reply:
@junrao Great point.. I have been thinking about the right solution since your comment. I have addressed the problem with the following changes to the patch:

- Add methods truncateTo() and truncateFullyAndStartAt() in Partition.scala. Both methods need to grab readLock of leaderIsrUpdateLock before truncating the log.

- Add method appendRecordsToFutureReplica() in Partition.scala. This method needs to grab inWriteLock of leaderIsrUpdateLock first. If LEO of the current replica >= LEO of the future replica, this method will append data to the future log. Otherwise, this method will throw OffsetOutOfRangeException. This method ensures that, if the current replica is truncated after ReplicaAlterDirThread fetches from the current replica but before ReplicaAlterDirThread try to append data to the future replica, the data will not be appended.

- AbstractFetcherThread.processFetchRequest() will catch OffsetOutOfRangeException thrown from processPartitionData(). When this exception is caught, the future replica will be marked for truncation. And the future replica will be truncated based on the epoch when ReplicaAlterDirThread calls maybeTruncate() next time.

Does this solution sound OK?

Reply:
My solution has a flaw. Let me think about how to fix it..

Reply:
I was thinking that one way to do this is that if the current replica truncates, we just remove the partition from the ReplicaAlterDirThread and add it back again. This will force the initialization with proper truncation if needed.

Reply:
@junrao Ah, I didn't realize you have replied to this thread.

Yeah I have considered this option. My concern with this approach is that truncation usually happens at the Partition level or Log level, but the management of ReplicaAlterDirManager and ReplicaFetcherManager ideally should happen at the ReplicaManager level, which is higher than the level of Partition and Log. Usually the level of caller should be higher than the level of caller to avoid deadlock and to make the code logic simpler to maintain and develop in the long term.

In this specific case, if we add/remove partition from ReplicaAlterDirManager, a ReplicaFetcherThread may need to get the following lock if truncation happens: AbstractFetcherThread.partitionMapLock, Partition.leaderIsrUpdateLock, ReplicaManager.replicaStateChangeLock, ReplicaAlterDirManager.mapLock and ReplicaAlterDirThread.mapLock. This may cause deadlock if another KafkaRequestHandler attempts to get the ReplicaManager.replicaStateChangeLock and then Partition.leaderIsrUpdateLock.

Do you think the approach I suggested previously would work? Do you have concern with having this approach depend on the leader epoch and thus the message format?

Reply:
Regarding your concern with "lock the writes to the current replica while writing to the future replica", I think it probably won't hurt performance much. It will reduce the maximum throughput of writing to the current replica by at most half. But I assume that for most replica, the average throughput should be much less than 50% of the maximum throughput that a broker can write to a partition. Furthermore, even if the throughput of this partition is very close to the maximum throughput, we actually want to reduce the throughput that a broker can write to the current replica before the future replica catches up. Does this make sense?

We can continue discussion under the new batch of comment you provided. Thanks!